### PR TITLE
stagedsync: Fix logIndex reset and missing websocket notifications in parallel execution

### LINKED ISSUE #22106: `eth_getLogs` returns inconsistent log indexes
#### System information

I can't fill out much of this information at the moment as I don't have direct access to the nodes we are running. Will update with the information when I have it.

Erigon version: `./erigon --version`

OS & Version: Linux

Commit hash: 

Erigon Command (with flags/config):

```
[INFO] [06-29|20:35:30.118] Startup command                          cmd="erigon --datadir=<redacted-dir> --chain=gnosis --externalcl --http.addr=<redacted-ip> --http.vhosts=any --http.api=eth,debug,net,trace,web3,erigon,txpool --rpc.blockrange.limit=100000 --ws --nat=extip:<redacted-ip> --port=30303 --metrics --metrics.addr=<redacted-ip> --private.api.addr=<redacted-ip> --torrent.download.rate=1000mb --batchSize=4g --prune.mode=archive --exec.serial --authrpc.addr=<redacted-ip> --authrpc.port=8551 --authrpc.vhosts=*"
```

Consensus Layer:

Consensus Layer Command (with flags/config):

Chain/Network: gnosis chain

#### Expected behaviour

identical `eth_getLogs` should return the same events and log indexes (right after the block was mined or a long time after the block was mined)

#### Actual behaviour

To get the most up-to-date data we subscribe to new blocks using websockets. Whenever we get informed about a new block we execute `eth_getLogs` on that block and write the relevant events into a database.
Since switching to erigon v3.5.0 a few days ago the events in our DB don't have the correct log index anymore. When we compare the indexed data with `cast logs` it reports the same transactions but sometimes they have different log indexes.

Without any proof or digging deep into the code I suspect we are running into a race condition where we fetch the logs on a block that has not been fully processed yet and the results of the parallel workers might not have been consolidated correctly yet.

Since we discovered this we switched to running with `--exec.serial`. However, we are still getting logs like `[5/5 Execution] parallel executed` which suggests that parallel execution is still active. After digging into the code it seems like this flag only limits the worker count to 1 but if there is work stealing scheduling going on the race conditions I suspect might still manifest.
~~We'll check if the incorrect indexes disappear if we also set `EXEC3_PARALLEL=false` to hopefully truly revert back to the old execution engine.~~
Running with `EXEC3_PARALLEL=false` indeed resolves the issue.

#### Steps to reproduce the behaviour

Execute `eth_getLogs` immediately after a WS notification about a new block and compare that result with the same call 1 minute later.

### PR DESCRIPTION
**Pull request by Sahil-4555**
### Issue

When Erigon is doing parallel block execution (`EXEC3_PARALLEL=true`), the state snapshots can end at step boundaries which fall mid-block. So when we restart the node or resume sync from these snapshots, the execution stage needs to start/resume execution from the middle of the block (running only the remaining transactions of that block).

While finalizing the transaction receipts for this partial block execution, the code in `exec3_parallel.go` was checking `if txVersion.TxIndex > 0 && tx > 0` to fetch the previous transaction's receipt from the memory results map. But for the first transaction task of this new batch (where the local slice task index `tx` is `0`, but the actual block-level index `txVersion.TxIndex` is `> 0`), this check was failing and `prevReceipt` remained `nil`. Because of this, it called `CreateNextReceipt(nil)` and the starting log index and cumulative gas for that mid-block transaction got reset to `0`. This wrong log index got written directly to the database `ReceiptDomain` and caused inconsistent `logIndex` values when clients queried `eth_getLogs`.

Additionally, for partial blocks the websocket notification path was publishing tail-only receipts (only the resumed portion) via `RecentReceipts.Add`, which `eth_subscribe` log/receipt clients would receive as the complete block — diverging from `eth_getLogs` results.

### Fix

The main fix is in the parallel executor's receipt finalization loop in `exec3_parallel.go`. We changed the `finalize` function signature to accept `cumulativeGasUsed` and `firstLogIndex` directly instead of a `prevReceipt` pointer, and it now calls `CreateReceipt` instead of `CreateNextReceipt`. This matches how the serial executor handles receipt creation. We also branch explicitly on `tx == 0` (batch boundary) instead of checking `prevReceipt == nil`. When the first task in a batch has `TxIndex > 0` (meaning we are resuming mid-block), we query the temporal database using`rawtemporaldb.ReceiptAsOf` to fetch the previous transaction's cumulative gas used, cumulative blob gas used, and log index offset. We excluded `IsBlockEnd` tasks from this fallback since `finalizeSystemTx` never consumes `prevReceipt` anyway. Also added a two-step nil check on `be.finalizedResults[tx-1]` before accessing `.Receipt` since `finalizedResults` is a map and a missing key would return nil and cause a panic.

For the notification issue, we gated `RecentReceipts.Add` behind `!applyResult.isPartial` and moved it outside the block validation scope so that incomplete tail-only receipts are not published for partial blocks. We now use `applyResult.Txs` and `applyResult.Header` directly instead of fetching from the database. This is consistent with how the serial path gates behind `startTxIndex == 0` in `exec3_serial.go:421`.

For the blob gas issue, we added a `cumulativeBlobGasUsed` field to `execResult` and snapshot `be.blobGasUsed` into it at finalization time. The publish loop now uses this per-result snapshot instead of reading `be.blobGasUsed` directly, which could have been overwritten by a later transaction. We also initialize `be.blobGasUsed` from the database `cumBlobGasUsed` value at the resume boundary.

Added `TestParallelResumeBoundaryAndNotifications` unit test to pin theresume boundary receipt reconstruction (first task with `TxIndex > 0` at slice index 0 → correct offsets from DB) and the `isPartial` flag that controls notification skipping.

Closes: https://github.com/erigontech/erigon/issues/22106

### CODE REVIEW

File: execution/stagedsync/exec3_parallel.go [hunk 1]
```diff
@@ -569,7 +570,7 @@ func (pe *parallelExecutor) execImpl(ctx context.Context, execStage *StageState,
 						blockValidatorWaiter = newBlockValidator(applyResult.BlockGasUsed, applyResult.BlobGasUsed, checkReceipts, checkBloom, applyResult.Receipts,
 							lastHeader, b.Transactions(), pe.cfg.chainConfig, pe.logger)
 
-						if !applyResult.isPartial && !execStage.CurrentSyncCycle.IsInitialCycle {
+						if !execStage.CurrentSyncCycle.IsInitialCycle {
```
Comment (Sahil-4555) [inline execution/stagedsync/exec3_parallel.go]: Actually, `applyResult.isPartial` becomes true when the block execution started mid-block (meaning the block execution was split across batches and this batch is finishing it). But even if it is a partial block, the block itself is fully completed in this batch. 

If we keep the `!applyResult.isPartial` check, then we are skipping `RecentReceipts.Add` completely for these blocks. Because of this, the websocket newHeads and logs subscriptions will never get the log notifications for the transactions that were executed in this second batch. 

So by removing this `!applyResult.isPartial` check, we are making sure that whenever a block is completed, its receipts are added to the `RecentReceipts` cache, and the websocket subscribers get all the logs properly. It is also safe because `RecentReceipts.Add` already returns early if the receipts list is empty.

File: execution/stagedsync/exec3_parallel.go [hunk 2]
```diff
@@ -2481,9 +2482,21 @@ func (be *blockExecutor) nextResult(ctx context.Context, pe *parallelExecutor, r
...
+					}
+					if prevReceipt == nil {
+						cumGasUsed, _, logIndexAfterTx, err := rawtemporaldb.ReceiptAsOf(applyTx, txVersion.TxNum)
+						if err != nil {
+							return nil, err
+						}
+						prevReceipt = &types.Receipt{
+							CumulativeGasUsed:        cumGasUsed,
+							FirstLogIndexWithinBlock: logIndexAfterTx,
+						}
```
Comment (Sahil-4555) [inline execution/stagedsync/exec3_parallel.go]: during parallel block execution, when we validate and finalize the transactions, we fetch the previous transaction's receipt from the `be.finalizedResults[tx-1]` map. This works fine when everything is in memory.

But if the block is split across batches, the batch restarts mid-block. For the very first transaction task of this new batch (where local index `tx` is `0`, but actual block index `txVersion.TxIndex` is `> 0`), the previous transactions were already committed in the database in the previous batch. Since `tx` is `0`, we cannot fetch from `finalizedResults[tx-1]` and `prevReceipt` becomes `nil`.

To solve this, if `prevReceipt` is `nil` but we are at a mid-block transaction (`txVersion.TxIndex > 0`), we query the database using `rawtemporaldb.ReceiptAsOf` to get the cumulative gas used and log index of the previous transaction. We then build a dummy `prevReceipt` with these values so that the log index and cumulative gas calculations start correctly from the database values instead of resetting to `0`. This is the exact same fallback logic that the serial executor (`exec3_serial.go`) uses at batch boundaries.

File: execution/stagedsync/exec3_parallel.go [hunk 3]
```diff
@@ -569,7 +570,7 @@ func (pe *parallelExecutor) execImpl(ctx context.Context, execStage *StageState,
 						blockValidatorWaiter = newBlockValidator(applyResult.BlockGasUsed, applyResult.BlobGasUsed, checkReceipts, checkBloom, applyResult.Receipts,
 							lastHeader, b.Transactions(), pe.cfg.chainConfig, pe.logger)
 
-						if !applyResult.isPartial && !execStage.CurrentSyncCycle.IsInitialCycle {
+						if !execStage.CurrentSyncCycle.IsInitialCycle {
 							pe.cfg.notifications.RecentReceipts.Add(applyResult.Receipts, b.Transactions(), lastHeader)
 						}
 					}
```
[the remark this answers is not in the document]
Comment (Sahil-4555) [reply execution/stagedsync/exec3_parallel.go]: Yes, actually the outer block is checking `!isPartial`, so my inner change was getting skipped anyway when `isPartial` was true. 

I have updated the code now and moved the `RecentReceipts.Add` call outside of that outer check. I am using `applyResult.Txs` and `applyResult.Header` directly as you suggested. This is much cleaner and avoids the extra DB query too. Thanks!

File: execution/stagedsync/exec3_parallel.go [hunk 4]
```diff
@@ -2489,9 +2491,21 @@ func (be *blockExecutor) nextResult(ctx context.Context, pe *parallelExecutor, r
...
+					}
+					if prevReceipt == nil {
+						cumGasUsed, _, logIndexAfterTx, err := rawtemporaldb.ReceiptAsOf(applyTx, txVersion.TxNum)
+						if err != nil {
+							return nil, err
+						}
+						prevReceipt = &types.Receipt{
+							CumulativeGasUsed:        cumGasUsed,
+							FirstLogIndexWithinBlock: logIndexAfterTx,
+						}
```
[the remark this answers is not in the document]
Comment (Sahil-4555) [reply execution/stagedsync/exec3_parallel.go]: Actually, this is a very good catch! 

In Erigon v3, parallel execution only runs on databases with schema version `>= v1_1` where `ReceiptStoresFirstLogIdx` is always `false`. So in practice, this legacy case won't be triggered.

But to make sure there is no silent index corruption if someone somehow runs it on a legacy schema, I have added an explicit check. If `ReceiptStoresFirstLogIdx(applyTx)` is `true`, we immediately return an error: `"parallel execution is not supported on databases with legacy receipt domain schemas (ReceiptStoresFirstLogIdx == true)"`. This will block execution cleanly instead of writing wrong indexes to the DB.

**Comment by sudeepdino008** [comment]
> like when the batch limit is reached mid-block, or when we restart the node and resume sync

we don't "execute blocks incompletely". State snapshots can end at midblock however and so execution needs to resume from middle of the block.

**Comment by sudeepdino008** [comment]
one weird thing:
- prevReceipt is usually coming from current batch's execution, so it have logs.
- when we use prevReceipt to construct next receipt, we do `currentReceipt.FirstLogIndexWithinBlock = prevReceipt.FirstLogIndexWithinBlock + len(logs)`
- in edge case when we retrieve from db, receipt has no logs (rcache has it, receipt doesn't). Also, prevReceipt created with no logs. 
- Also, in this case, `logIndexAfterTx` retrieved from db is actually the index after adding logs...

but it works out in end
`prevReceipt.FirstLogIndexWithinBlock = logIndexAfterTx`.  <-wrong assignment
but
`currentReceipt.FirstLogIndexWithinBlock = prevReceipt.FirstLogIndexWithinBlock + len(logs)`

logs is empty, so currentReceipt.FirstLogIndexWithinBlock gets set with the right value..

I'd suggest adding a comment about this

---


- exec3_serial did it properly, by re-executing the prev tx and getting logs etc. But seems we don't need it. 
- I wanna get rid of the receipt version check, but bloatnet still has it. I think bloatnet is dead, but if it isn't - i'll regen the receipt and get rid of this check.

**Review by sudeepdino008 — APPROVED** [review]

**Comment by Sahil-4555** [comment]
You are totally right that `prevReceipt.FirstLogIndexWithinBlock = logIndexAfterTx` isn’t technically the *previous* starting index, but since we set `Logs = nil` (so `len(Logs) == 0`) the calculation `logIndexAfterTx + 0` gives the starting index for the next transaction. This allows us to reuse `CreateNextReceipt` directly, without extra DB fetching or re-executing the previous transaction.

I’ve added a clear code comment explaining this mathematical trick. Thanks for catching this!

File: execution/stagedsync/exec3_parallel.go [hunk 5]
```diff
@@ -2494,9 +2496,31 @@ func (be *blockExecutor) nextResult(ctx context.Context, pe *parallelExecutor, r
...
+					}
+					if prevReceipt == nil {
+						if rawtemporaldb.ReceiptStoresFirstLogIdx(applyTx) {
+							return nil, fmt.Errorf("parallel execution is not supported on databases with legacy receipt domain schemas (ReceiptStoresFirstLogIdx == true)")
+						}
+						cumGasUsed, _, logIndexAfterTx, err := rawtemporaldb.ReceiptAsOf(applyTx, txVersion.TxNum)
+						if err != nil {
+							return nil, err
+						}
+						// Actually, when we reconstruct prevReceipt from the DB, it does not have any logs.
```
Comment (lupin012) [inline execution/stagedsync/exec3_parallel.go]: Here should be initialized also be.blobGasUsed = cumBlobGasUsed

File: execution/stagedsync/exec3_parallel.go [hunk 6]
```diff
@@ -2494,9 +2496,31 @@ func (be *blockExecutor) nextResult(ctx context.Context, pe *parallelExecutor, r
...
+					if tx > 0 {
+						if prev := be.finalizedResults[tx-1]; prev != nil {
+							prevReceipt = prev.Receipt
+						}
+					}
+					if prevReceipt == nil {
+						if rawtemporaldb.ReceiptStoresFirstLogIdx(applyTx) {
+							return nil, fmt.Errorf("parallel execution is not supported on databases with legacy receipt domain schemas (ReceiptStoresFirstLogIdx == true)")
+						}
+						cumGasUsed, _, logIndexAfterTx, err := rawtemporaldb.ReceiptAsOf(applyTx, txVersion.TxNum)
```
Comment (lupin012) [inline execution/stagedsync/exec3_parallel.go]: cumGasUsed, cumBlobGasUsed , logIndexAfterTx, err := ....

**Review by lupin012 — APPROVED** [review]

**Review by yperbasis — CHANGES_REQUESTED** [review]
The receipt-index reconstruction itself checks out, but a few things need fixing before merge:

1. **Partial blocks now notify with tail-only receipts** (`exec3_parallel.go:579`). For an `isPartial` block, `blockResult.Receipts` contains only the resumed tail (built from `be.results`; the task producer skips the prefix), while `Txs`/`Header` are the full block's. `RecentReceipts.Add` stores this as the block's complete set, so `eth_subscribe` log/receipt clients get the tail presented as the whole block — diverging from `eth_getLogs`, replaying `Removed` on unwind for the tail only, and bypassing the `!isPartial`-gated receipts-root/bloom validation. Please either reconstruct the prefix (`receipts.DerivePriorReceipts`, as the serial path does for `Finalize`) or keep skipping partial blocks like serial does (`startTxIndex == 0`, exec3_serial.go:421); publishing an incomplete set is worse than either.

2. **The resume-at-block-end offset still gets zero notifications** (`exec3_parallel.go:2505`). If the boundary falls between the last user tx and the block-end txNum, the resumed executor's only task is block-end: `Receipts` is empty, `Add` early-returns, and the completed block emits newHeads but no logs/receipts — the #22106 symptom remains. The fallback also runs for block-end tasks even though `finalizeSystemTx` never consumes `prevReceipt`; please exclude `IsBlockEnd` tasks.

3. **The legacy-schema check can never fire** (`exec3_parallel.go:2506`). `ReceiptStoresFirstLogIdx` tests `CurrentDomainVersion < v1.1`, but the exec stage only ever sees `{3,0}` (schema default) or `{1,1}` (`AdjustReceiptCurrentVersionIfNeeded` pins legacy datadirs to exactly v1.1), so the guard is dead code — while 3.0-written data tails still pass through and get exactly the shifted indexes the error message claims to prevent. The underlying problem is the predicate threshold (arguably `Less(V2_0)`; dates to #16677) — better to fix that than ship a check that suggests protection it doesn't provide.

4. **`be.blobGasUsed = cumBlobGasUsed` seeds from a slot the parallel executor writes per-tx, not cumulative** (`exec3_parallel.go:2525`): the publish loop passes `tx.GetBlobGas()` as `ApplyTxIndexes`' `cummulativeBlobGas` arg (exec3_parallel.go:2740), unlike serial which passes the running cumulative (exec3_serial.go:556). On parallel-synced datadirs the seed is the previous tx's own blob gas (0 for non-blob txs). Latent today because the validator is skipped for partial blocks, but wrong — fix the write side to store the running cumulative, or drop the seed.

5. **Shape**: branch explicitly on `tx == 0` instead of `prevReceipt == nil` (the nil case is otherwise impossible; if that invariant ever breaks, this silently reconstructs from pre-batch DB state instead of failing loudly), and consider passing the two scalars through `finalize` into the existing `CreateReceipt` like serial does (exec3_serial.go:452-461) — `finalize` has a single call site — rather than a synthetic receipt whose `FirstLogIndexWithinBlock` holds the *next* tx's index. That also removes the need for the 7-line comment (which restates `CreateNextReceipt`'s math and already drifts from it); the `// Initialize cumulative blob gas...` comment just restates the assignment and can go.

6. Per repo guidelines (CLAUDE.md, TDD for bug fixes), please add a test pinning the resume boundary (first task with `TxIndex > 0` at slice index 0 → reconstructed receipt values) and the notification behavior — nothing currently prevents a regression of the #22106 corruption.

**Comment by Sahil-4555** [comment]
Hi @yperbasis,

Thanks for the detailed feedback. I have gone through all your points and done the needful changes. Please find the updates below:

1. **Partial block notifications**: We are now skipping notifications for partial blocks completely, same like how the serial execution path is doing. We check `!applyResult.isPartial` before calling `RecentReceipts.Add` now.
2. **Excluding block-end**: Excluded the `IsBlockEnd` tasks from the fallback database query since finalize system tx does not need it anyway.
3. **Dead code removal**: Removed the legacy-schema check completely as it was dead code and not firing.
4. **Blob gas seed**: We initialized `be.blobGasUsed` from the db `cumBlobGasUsed` as per the comment.
5. **Code shape**: We changed the branching logic to check `tx > 0` directly instead of checking if receipt is nil. Also, we are passing the offset values directly into `CreateReceipt` via `finalize` now, and cleaned up those long comments.
6. **Unit test**: Added `TestParallelResumeBoundaryAndNotifications` unit test to verify this boundary fallback and partial block notification skipping. All tests are passing fine.

Please review and let me know if it looks good now.

File: execution/stagedsync/exec3_parallel.go [hunk 7]
```diff
@@ -573,9 +574,10 @@ func (pe *parallelExecutor) execImpl(ctx context.Context, execStage *StageState,
...
 							lastHeader, b.Transactions(), pe.cfg.chainConfig, pe.logger)
 
-						if !applyResult.isPartial && !execStage.CurrentSyncCycle.IsInitialCycle {
-							pe.cfg.notifications.RecentReceipts.Add(applyResult.Receipts, b.Transactions(), lastHeader)
-						}
+					}
+
+					if applyResult.BlockNum > 0 && !applyResult.isPartial && !execStage.CurrentSyncCycle.IsInitialCycle && applyResult.Header != nil {
+						pe.cfg.notifications.RecentReceipts.Add(applyResult.Receipts, applyResult.Txs, applyResult.Header)
 					}
```
[the remark this answers is not in the document]
Comment (Sahil-4555) [reply execution/stagedsync/exec3_parallel.go:581]: Thanks for the heads-up, but this is intentional. When resuming a block mid-batch, blockResult. Receipts only contains the tail receipts, prefix txs were never run in this batch. Publishing these as the full block’s receipts would give eth_subscribe clients an incomplete set, which is exactly what @yperbasis asked us to avoid. The serial path does the same (exec3_serial.go:421 gates behind startTxIndex == 0). eth_getLogs still works fine as it reads from the temporal DB.

File: execution/stagedsync/exec3_parallel.go [hunk 8]
```diff
@@ -2493,10 +2498,22 @@ func (be *blockExecutor) nextResult(ctx context.Context, pe *parallelExecutor, r
...
-						prevReceipt = prev.Receipt
+				var cumulativeGasUsed uint64
+				var firstLogIndex uint32
+				if txVersion.TxIndex > 0 && !txTask.IsBlockEnd() {
+					if tx > 0 && be.tasks[tx-1].Task.Version().TxIndex >= 0 {
+						if prev := be.finalizedResults[tx-1].Receipt; prev != nil {
+							cumulativeGasUsed = prev.CumulativeGasUsed
+							firstLogIndex = prev.FirstLogIndexWithinBlock + uint32(len(prev.Logs))
+						}
+					} else {
```
[the remark this answers is not in the document]
Comment (Sahil-4555) [reply execution/stagedsync/exec3_parallel.go:2509]: applied the two-step nil guard. Thanks!

File: execution/stagedsync/exec3_parallel_test.go [hunk 9]
```diff
@@ -1349,3 +1353,131 @@ func BenchmarkDexScenarioWithMetadata(b *testing.B) {
...
+			cfg: ExecuteBlockCfg{
+				chainConfig: chainSpec.Config,
+				db:          db,
+				engine:      ethash.NewFaker(),
+			},
+			doms:   domains,
+			rs:     state.NewStateV3Buffered(state.NewStateV3(domains, false, logger)),
+			logger: logger,
+		},
+		workerCount: runtime.NumCPU() - 1,
```
[the remark this answers is not in the document]
Comment (Sahil-4555) [reply execution/stagedsync/exec3_parallel_test.go:1432]: Good point. But all current tests in this file (533, 656, 713 lines) use the same runtime. NumCPU() - 1 pattern, no max guard. It would be inconsistent to fix only our test: this is a pre-existing pattern and out of scope 
for this PR . Happy to handle it in a separate cleanup if you like.

File: execution/stagedsync/exec3_parallel_test.go [hunk 10]
```diff
@@ -1349,3 +1353,131 @@ func BenchmarkDexScenarioWithMetadata(b *testing.B) {
 
 	testExecutorCombWithMetadata(b, totalTxs, numReads, numWrites, numNonIO, taskRunner, logger)
 }
+
+func TestParallelResumeBoundaryAndNotifications(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (Sahil-4555) [reply execution/stagedsync/exec3_parallel_test.go:1357]: We do assert notification behaviour - line 1482 does check res.isPartial == true, which is the exact flag that gates RecentReceipts. Add at exec3_parallel.go:580. The @yperbasis requested isPartial pinning. RecentReceipts. Add dispatch would involve mocking the notifications subsystem, which seems like overkill here.

File: execution/stagedsync/exec3_parallel.go [hunk 11]
```diff
@@ -2493,10 +2498,22 @@ func (be *blockExecutor) nextResult(ctx context.Context, pe *parallelExecutor, r
...
+						cumGasUsed, cumBlobGasUsed, logIndexAfterTx, err := rawtemporaldb.ReceiptAsOf(applyTx, txVersion.TxNum)
+						if err != nil {
+							return nil, err
+						}
+						cumulativeGasUsed = cumGasUsed
+						firstLogIndex = logIndexAfterTx
+						be.blobGasUsed = cumBlobGasUsed
```
[the remark this answers is not in the document]
Comment (Sahil-4555) [reply execution/stagedsync/exec3_parallel.go:2516]: @yperbasis asked to remove this check explicitly in his request changes - the ReceiptStoresFirstLogIdx predicate has a wrong threshold ( v1.1 instead of  v2.0) and never fires for schemas the exec stage actually sees ({3,0} or {1,1}) putting it back in would be dead code that implies protection it doesn't provide. The fix itself is a different story (dates back to #16677).

File: execution/stagedsync/exec3_parallel.go [hunk 12]
```diff
@@ -2493,10 +2498,22 @@ func (be *blockExecutor) nextResult(ctx context.Context, pe *parallelExecutor, r
...
+							firstLogIndex = prevRes.Receipt.FirstLogIndexWithinBlock + uint32(len(prevRes.Receipt.Logs))
+						}
+					} else {
+						cumGasUsed, cumBlobGasUsed, logIndexAfterTx, err := rawtemporaldb.ReceiptAsOf(applyTx, txVersion.TxNum)
+						if err != nil {
+							return nil, err
+						}
+						cumulativeGasUsed = cumGasUsed
+						firstLogIndex = logIndexAfterTx
+						be.blobGasUsed = cumBlobGasUsed
```
[the remark this answers is not in the document]
Comment (Sahil-4555) [reply execution/stagedsync/exec3_parallel.go:2516]: Updated the PR description - it no longer mentions the legacy schema check. We removed that guard per @yperbasis review since the predicate threshold is wrong and never fires. The new description reflects the current behavior.

File: execution/stagedsync/exec3_parallel.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (Sahil-4555) [reply execution/stagedsync/exec3_parallel.go:581]: Updated the PR description - it now clearly states that partial blocks skip notifications. This is intentional to avoid publishing incomplete tail-only receipts, consistent with the serial path

**Review by yperbasis — APPROVED** [review]
